### ISSUE TITLE
Implement optimized closed-form gradients for Kalman Filter

### ISSUE THREAD

**Issue by jessegrabowski** [pymc-devs/pymc-experimental#332]
I'm quite interested in the results of [this paper](https://arxiv.org/pdf/2303.16846.pdf). The authors  derive closed-form gradients for backprop through Kalman Filters. Specifically equations 28-31.

They report a 38x speedup over autodiff gradients from PyTorch. I suspect (with no evidence) that the gradient computations are where the default PyMC sampler really fall down, so this might even make non-JAX sampling of SS models palatable.

**Comment by victorgarcia98** [comment]
Hey @jessegrabowski, 

Cool paper!
I would like to contribute towards implementing (or at least trying) this gradients. To start, I can try to reproduce the paper and then I'll need some help to know where to integrate these gradients.

What do you think? 😄

**Comment by jessegrabowski** [comment]
Yeah that's how I'd proceed. You have closed form gradients from the current implementation, so basically you can just try to implement their equations 28-31. For a test model use a local level model. I recommend this because it doesn't have any parameters (except for the innovation covariance, and measurement error if included)

A big contribution to just getting the ball rolling would also just be to work out the notation differences between what this paper uses and what we use (we follow Durbin and Koopman Time Series Analysis by Statespace Methods).

**Comment by victorgarcia98** [comment]
Awesome! Sounds like I plan :D

I'll create a table comparing the notations!

**Comment by JeanVanDyk** [comment]
# Hi all

This post is a preliminary draft. I will likely revise and edit it further as the notations and concepts become clearer and better defined.

Here, I will try to synthesize the article and explain what these closed-form expressions are, which supposedly lead to a significant speed-up in the gradient computation time. I will use the notations from *"Durbin and Koopman, Time Series Analysis by State Space Methods"*. (To be fully honest, I mostly relied on the notations from *"Statistical algorithms for models in state space using SsfPack 2.2"*, which I hope are equivalent.)

First, to establish a common ground, let’s define the transition equations used in the Kalman filter:
 
$$
\begin{align}
a_{t+1} &= c_t + T_t a_t + H_t \epsilon_t \qquad \text{where } a_1 \sim \mathcal{N}(a, P) \\
y_t &= d_t + Z_t a_t + G_t \epsilon_t \qquad \text{where } \epsilon_t \sim \mathcal{N}(0, I)
\end{align}
$$

### Parameter definitions

For clarity, here are some key parameters:

#### State equation parameters

- $a_t$ : state vector at time t.
- $T_t$ : state transition matrix at time t.
- $c_t$ : intercept vector at that time t.
- $H_t$ : selection matrix at time t.

#### Observation equation parameters

- $y_t$ : observation at time t.
- $Z_t$ : the design matrix at time t.
- $d_t$ : intercept vector at that time t.
- $G_t$ : observation error matrix at time t.

### Update and Prediction Equations

We now define the notation used for the update and prediction steps.

#### Update equations

$$
\begin{align}
v_t &= y_t - d_t - Z_t \hat{a}\_{t} \\
F_t &= Z_t\hat{P}\_tZ^T_t + H_t \\
K_t &= \hat{P}\_tZ^T_tF_t^{-1} \\
a_{t+1} &=  \hat{a}\_{t} + K_t v_t \\
P_{t+1} &= (I - K_t Z_t) \hat{P}_t (I - K_t Z_t)^T + K_t H_t K^T_t
\end{align}
$$

#### Prediction equations

$$
\begin{align}
\hat{a}\_{t+1} &= c_t + T_t a\_{t+1} \\
\hat{P}_{t+1} &= T_t P\_{t+1} T^T_t + R_t Q_t R^T_t
\end{align}
$$

### Computing the Loss

The usual way to compute the loss (log-likelihood contribution) at time \(t\) is:

$$
\begin{align}
ll_t &= -0.5 (log(2 \pi) + log(|det(F_t)|) +v_t^T F_t^{-1} v_t)
\end{align}
$$

## PyMC’s Default Gradient Computation

From my understanding, please feel free to correct me if I’m mistaken, PyMC’s `sample()` method, whether relying on the default NUTS or its JAX implementation, uses automatic differentiation to compute gradients. These gradients are then used in the sampling process as a guide to indicate where the next sample should be drawn in the parameter space.

In the worst-case scenario, this involves calculating a Jacobian matrix of size ( d x d ), where d is the dimension of the state or parameter vector. Naively, computing this Jacobian has a complexity of approximately $\mathcal{O}(d^4)$, due to the tensor product operations on square matrices.

More specifically, when differentiating matrix operations such as the inverse of a matrix M, the Jacobian involves terms like:

$$
\frac{\partial M^{-1}}{\partial M} = - M^{-1} \otimes M^{-1}
$$

However, in a temporal or recurrent setting where gradients propagate backward through N time steps, the Jacobians are multiplied across these steps. This results in an overall computational complexity of $\mathcal{O}(N d^5)$, as mentioned in the referenced article.

## Article's Results

The article demonstrates that it is possible to obtain closed-form expressions for the partial derivatives of the loss with respect to all model parameters. By doing so, we avoid the need to compute high-dimensional Jacobians explicitly, which are typically required when using automatic differentiation.

This closed-form approach significantly reduces computational complexity. Instead of the worst-case scenario of $\mathcal{O}(N d^5)$, we now deal with a much more tractable $\mathcal{O}(N d^3)$.

$$
\begin{align}
\frac{\partial ll_t}{\partial P_t} &= 0 \\
\frac{\partial ll_t}{\partial a_t} &= 0 \\
\frac{\partial \hat{ll}\_t}{\partial \hat{P}\_t} &= Z_t^T F_t^{-1} Z_t -  Z_t^T F_t^{-1} z_t z_t^T F_t^-1 Z_t \\
\frac{\partial \hat{ll}\_t}{\partial \hat{a}\_t} &= - 2 Z_t^T F_t^{-1} y_t \\
\frac{\partial \hat{ll}\_t}{\partial H\_t} &= F_t^{-1} - F_t^{-1} z_t z_t^T F_t^-1 \\
\frac{\partial \hat{ll}\_t}{\partial \hat{a}\_t} &= - 2 F_t^{-1} z_t \\
&\text{where } z_t = y_t -Z_t \hat{a_t}
\end{align}
$$
